pkg/controller/machine_util.go

```diff
@@ -301,87 +347,161 @@ func (c *controller) syncMachineNodeTemplates(machine *v1alpha1.Machine) error {
+			nLabels[mKey] = mValue
+			toBeUpdated = true
+		}
 	}
 
-	return updateNeeded
+	return toBeUpdated
 }
 
-// SyncMachineTaints syncs the annotations of the machine with node-objects.
+type taintKeyEffect struct {
+	// Required. The taint key to be applied to a node.
+	Key string
+	// Valid effects are NoSchedule, PreferNoSchedule and NoExecute.
+	Effect v1.TaintEffect
+}
+
+// SyncMachineTaints syncs the taints of the machine with node-objects.
 // It returns true if update is needed else false.
-func SyncMachineTaints(machine *v1alpha1.Machine, node *v1.Node) bool {
-	mTaintsCopy, nTaintsCopy := machine.Spec.NodeTemplateSpec.Spec.Taints, node.Spec.Taints
-	updateNeeded := false
-
-	for i := range mTaintsCopy {
-		elementFound := false
-		for j := range nTaintsCopy {
-			if mTaintsCopy[i] == nTaintsCopy[j] {
-				elementFound = true
-				break
-			} else if mTaintsCopy[i].Key == nTaintsCopy[j].Key {
-				elementFound, updateNeeded = true, true
-				nTaintsCopy[j] = mTaintsCopy[i]
-				break
-			}
-		}
-		if elementFound == false {
-			nTaintsCopy = append(nTaintsCopy, mTaintsCopy[i])
-			updateNeeded = true
+func SyncMachineTaints(
+	machine *v1alpha1.Machine,
+	node *v1.Node,
+	lastAppliedTaints []v1.Taint,
+) bool {
+	toBeUpdated := false
+	mTaints, nTaints := machine.Spec.NodeTemplateSpec.Spec.Taints, node.Spec.Taints
+	mTaintsMap := make(map[taintKeyEffect]*v1.Taint, 0)
+	nTaintsMap := make(map[taintKeyEffect]*v1.Taint, 0)
+
+	// Initialize node taints if nil
+	if nTaints == nil {
+		nTaints = []v1.Taint{}
+	}
+	// Intialize machine taints to empty map if nil
+	if mTaints == nil {
+		mTaints = emptyTaintList
+	}
+
+	// Convert the slice of taints to map of taint [key, effect] = Taint
+	// Helps with indexed searching
+	for i := range mTaints {
+		mTaint := &mTaints[i]
+		taintKE := taintKeyEffect{
+			Key:    mTaint.Key,
+			Effect: mTaint.Effect,
 		}
+		mTaintsMap[taintKE] = mTaint
 	}
-	if updateNeeded {
-		node.Spec.Taints = nTaintsCopy
+	for i := range nTaints {
+		nTaint := &nTaints[i]
+		taintKE := taintKeyEffect{
+			Key:    nTaint.Key,
+			Effect: nTaint.Effect,
+		}
+		nTaintsMap[taintKE] = nTaint
+	}
+
+	// Delete taints that existed on the machine object in the last update but anymore
+	for _, lastAppliedTaint := range lastAppliedTaints {
+
+		lastAppliedKE := taintKeyEffect{
+			Key:    lastAppliedTaint.Key,
+			Effect: lastAppliedTaint.Effect,
+		}
+
+		if _, exists := mTaintsMap[lastAppliedKE]; !exists {
+			delete(nTaintsMap, lastAppliedKE)
+			toBeUpdated = true
+		}
+	}
+
+	// Add any taints that exists in the machine object but not on the node object
+	for mKE, mV := range mTaintsMap {
+		if nV, exists := nTaintsMap[mKE]; !exists || *nV != *mV {
+			nTaintsMap[mKE] = mV
+			toBeUpdated = true
+		}
+	}
+
+	if toBeUpdated {
+		// Convert the map of taints to slice of taints
+		nTaints = []v1.Taint{}
+		for _, nV := range nTaintsMap {
+			nTaints = append(nTaints, *nV)
+		}
```

Comment:
Instead of 
```go
		nTaints = []v1.Taint{}
		for _, nV := range nTaintsMap {
			nTaints = append(nTaints, *nV)
		}
```

How about 

```go
		nTaints = make([]v1.Taint, len(nTaintsMap))
		i := 0
		for _, nV := range nTaintsMap {
			nTaints[i] = *nV
			i++
		}
```

Reply:
Done. Refer - https://github.com/gardener/machine-controller-manager/pull/268/commits/2d5030db1b9f88a560ddc8089446c9727145d135